specification/sections/01-flag-evaluation.md

```diff
@@ -31,6 +33,20 @@ OpenFeature.setProvider(new MyProvider());
 
 See [provider](./02-providers.md) for details.
 
+#### Requirement 1.1.2.2
```

Comment:
What happens if either `initialize` or `shutdown` fail for some reason?

Reply:
Good question.

I think the provider stays in a bad state (assuming the initialize is necessary for it to function properly) and probably throws during evaluation, and so the SDK will default for all evaluations.

Long term, my thinking is a failed initialize would fire the `ProviderEvents.ERROR` event as well, which would allow application authors to "do something" when this happens by attaching an event handler. I think that even in this case, the provider would still default until some corrective action is taken.

Reply:
I don't think I did a great job explaining my thinking here, so I'll try again :sweat_smile: 

I think that a lot of the things that can "go wrong" during the initialization in the `setProvider` can go wrong outside that context too. Persistent connections can be lost, APIs can become unavailable, etc. Therefor, we already have to handle these situations "out-of-band" of the `setProvider` function anyway, and I think using events is the best approach for this. For this reason, I think it's best to keep things simple here, and basically just log warnings if things go wrong in the `shutdown/initialization` within the `provider mutator`. The provider may attempt to re-establish whatever state is needed to work correctly, or give up if the error is irrecoverable. If it throws, the SDK will default.

That's my line of thinking at the moment.

Reply:
I think the `shutdown` is typically only called when the process exists so I think throwing an event at such a moment would not be really helpful. As don't think you can really act on it at least in Node.js. For example, if you get a SIGKILL signal or something similar.

I assume you are talking when you want to call `shutdown` during the life time of the application. I have to admit I personally haven't found a use case for switching providers in a running application but I think logging a warning works to notify that something went wrong and needs attention.

Reply:
> I think the `shutdown` is typically only called when the process exists so I think throwing an event at such a moment would not be really helpful. As don't think you can really act on it at least in Node.js. For example, if you get a SIGKILL signal or something similar.
> 
> I assume you are talking when you want to call `shutdown` during the life time of the application. I have to admit I personally haven't found a use case for switching providers in a running application but I think logging a warning works to notify that something went wrong and needs attention.

I think we agree. In this model here are 2 places a provider's shutdown is called:

1) when switching away from that provider (a new provider is set)
2) when the API's shutdown is called (this would be used when the application is terminating, as you mention)

Events don't help much in either case for `shutdown`. For errors during/after `initialize`, I think events are important. Events could be important to signal to the `application author` that a connection was lost or some fundamental state of the provider has changed, even long after initialization.

Reply:
So I think this proposal may be blocked until our events spec is landed.

Reading this: https://github.com/open-feature/spec/pull/171/files#diff-daf30d840f7e7b1752899ac2c4b268af214e6c8e7058549c8d4218abd2e86ba3

It would suggest that event handlers are registered on clients.  This would mean our code would look something like:

```java
OpenFeature api = OpenFeature.getInstance();
Client client = api.getClient('client-name')
client.addHook(ProviderEvent.ERROR, ....)
api.setProvider(new MyProvider())
```

Having to get a client before you've registered a provider seems kinda odd to me?

Reply:
@toddbaert I was thinking that we'd close out the events spec first, then this one.. not that we'd combine them.

If you expect someone to respond to a provider being in an error state with a log or metric or something, they will need to know about it. What I was attempting to show with the above block of code is that provider registration happens on the API level, but handling the errors that are generated happens at the Client level. This seems incongruous.

Reply:
> What I was attempting to show with the above block of code is that provider registration happens on the API level, but handling the errors that are generated happens at the Client level. This seems incongruous.

At first I thought this was an issue as well, but I think that this is a separation we want, TBH. The code setting the provider likely can't really react to such errors, I think it makes sense to have "client type" code react to it... I don't think that (for example) an alternative provider is going to be set because of an error. I think it's more realistic that areas of the app consuming feature flags (using `client` objects) would take some action or revert to some high-level default.

Reply:
@justinabrahms another possibility would be to add global events as well, and perhaps scope them with the same scope identifiers you're proposing in your multi-provider proposal. I'm not convinced this is a better way to go at this point though.

Reply:
It seems like we should have some mechanism to detect if there are errors during initialization without requiring a client. But if there is broad disagreement w/ that approach then don't let this be a blocker.

Reply:
@justinabrahms 

> It seems like we should have some mechanism to detect if there are errors during initialization without requiring a client.

I spent some time thinking about this. I think that in most real-world production cases, it's the [`application author`](https://openfeature.dev/specification/glossary#application-author) doing the feature evaluation and dealing with the client abstraction that needs to handle errors (loss of connection etc). The author of that particular part of the application knows how best to handle errors in that area.

That said, there's 2 use-cases I can think of where an [`application integrator`](https://openfeature.dev/specification/glossary#application-integrator) would need similar error signals:

- when initially configuring a provider, it might be tedious to have to get a client to troubleshoot connection issues, authentication issues, etc
- for global monitoring/telemetry purposes (sending a message to sentry or similar, for instance)

So I think I see your point. I have two ideas, which are not mutually exclusive:

1) cause `setProvider` to throw after a (configurable?) timeout, if there's an error in `initialize()`

This would at least help with the first problem. It doesn't really help too much with the second, since setting the provider is a one-time thing, and subsequent errors with the provider aren't captured without attaching a client. I think this would be easy to implement, and it wouldn't violate our "don't throw" rule since that doesn't apply to configuration. Practically, this might constitute a breaking change for JS implementations, since errors here would be associated with asynchronous code, meaning the `setProvider` would need to return a promise.

2) add "global" events. 

These would be similar to the client events, but not scoped to a single provider, wouldn't require the creation of a client. I think we still want most events to be associated with a client (expecially because in a mutli-provider scenario, we need a way to scope events). I think global events could be useful for adding handlers for monitoring the system in it's entirety, and not so much application-level concerns like flag value changes. Put another way, global events would be useful to  [`application integrators`](https://openfeature.dev/specification/glossary#application-integrator) and client events would be useful to [`application authors`]([`application authors`](https://openfeature.dev/specification/glossary#application-author)). We may want to only support error event types in global events.

Reply:
Global events could be annotated with the provider they reference. This would allow folks to tie into all events at once, which would be useful for the two use-cases mentioned above.

I agree that this is an application integrator concern. I don't have a preference between the two options you have. 2 seems more powerful. 1 seems easier to implement (in java).

Reply:
I also see the benefit of **global events** and IMO is the right way to support events. 

When an application starts, there is a setup phase where runtime configurations are set. I see Provider registration to be part of such a setup phase. And the client can be derived at business logic which can happen after some time (seconds to minutes depending on the usage) of the startup/setup phase. And it could be too late to detect events/failures at this point. Consider a K8s app with a startup probe. A developer can decide readiness to be set only if provider initialization is successful (aka `PROVIDER_READY`).

Reply:
_Not sure to which discussion to put this:_
Requirement 2.4.4 describes that transient errors can lead to the status flag changing between `ERROR` and `READY`. 
Should we specify that the change of the status Flag **MUST** lead to an event dispatched? (E.g. PROVIDER_READY and PROVIDER_ERROR)
Because if we only specify that there must be a (global) event on an error in `initialize()`, application integrators and application authors would get an inconsistent view on the state if the error is transient and status changes back to ready.

Reply:
> _Not sure to which discussion to put this:_ Requirement 2.4.4 describes that transient errors can lead to the status flag changing between `ERROR` and `READY`. Should we specify that the change of the status Flag **MUST** lead to an event dispatched? (E.g. PROVIDER_READY and PROVIDER_ERROR) Because if we only specify that there must be a (global) event on an error in `initialize()`, application integrators and application authors would get an inconsistent view on the state if the error is transient and status changes back to ready.

I agree in principle, but I think we'd have to put that in the events spec. Like I said in slack, these subjects are becoming a bit intertwined, but I'd like to keep the PRs isolated. I think when we release a new spec version, it should contain both init/shudown and events.

Reply:
Okay makes sense, so if we still want to include that, should it be added to the other PR (which is then the same problem, or what do we do to have it in the spec later @toddbaert?

Reply:
> Okay makes sense, so if we still want to include that, should it be added to the other PR (which is then the same problem that we are talking about something not present there), or what do we do to have it in the spec later @toddbaert? Because it also feels kind of wrong to add it to that PR and talk about `initialize()` which is not present there.

`Initiailize()` is present there, that PR is based on this one. I'm just trying to keep discussion scoped - the feedback we've had before is to try to keep PRs as on-topic as possible